### ISSUE TITLE
Efficient way to count number of isolated samples on a tree?

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/tskit#362]
I'm not sure if this deserves an issue, but there isn't a tskit mailing list yet (are we getting to the point where there should be?), and I can close if it's obvious.

Now that we indicate missing data using [isolated nodes](https://tskit.readthedocs.io/en/latest/data-model.html#sec-data-model-missing-data), is there an efficient way of finding the number of non-missing samples in a `Tree`? I could iterate through `Tree.roots()` checking which exist in `ts.samples()` and which don't have children, but according to the docs that operation is O(num_roots), so if there is lots of missing data that might be problematic. Alternatively I could do `[s for s in ts.samples() if tree.parent(s) == tskit.NULL and len(tree.children(s))==0]` but that might be tedious for large sample sizes. Is there an obvious method that I'm missing?

**Comment by petrelharp** [comment]
Well,
```
t = ts.first
sum([t.parent(n) is tskit.NULL for n in ts.samples()])
```
should get it reasonbly efficiently, but you're right, this'd be a good thing to have. Really, though, we should come up with a list of things operations we'd like to do about missing data, to have a consistent API about it?

**Comment by hyanwong** [comment]
Thanks @petrelharp. I agree re deeper thinking about what we want for simulations with missing data. We are starting to hitting these issues when using truncated samples (see https://github.com/tskit-dev/tskit/issues/361#issuecomment-529413956).

Re this specific problem, I wondered if iterating over all the samples for each tree would be tedious for large sample sizes (and for the record I think you probably want to check whether the sample has children too, before assuming it is isolated). I guess I could keep a running total by using edge diffs and working out which nodes are mentioned as parents or children. Something like

```
samples = set(ts.samples())
curr_samples = set()
for i, ((_, e_out, e_in), tree) in enumerate(
        zip(ts.edge_diffs(), ts.trees())):
    curr_samples.difference_update([e.parent for e in e_out if e.parent in samples])
    curr_samples.difference_update([e.child for e in e_out if e.child in samples])
    curr_samples.update([e.parent for e in e_in if e.parent in samples])
    curr_samples.update([e.child for e in e_in if e.child in samples])
    print("current missing samples are", samples-curr_samples)
```

**Comment by petrelharp** [comment]
Come to think of it, we could almost do this as a Node statistic.

**Comment by jeromekelleher** [comment]
I don't think we can do any better than iterating over the roots, and checking to see if they are isolated samples. I think it's reasonable to assume that number of missing samples << total number of samples for now. I don't think it'll end up being that expensive in practise.

We can add a ``num_isolated_samples`` property (or something similar) which does this  to the Tree class to make this more convenient. If it turns out that this every becomes a performance bottleneck, we can make it more efficient by adding an option to ``TreeSequence.trees()`` so that it is computed in C while we are doing the main iteration algorithm.

**Comment by hyanwong** [comment]
I probably misunderstood. I took the `O(n_roots)` comment in the docs to mean that, if we had a large number of roots in the tree (say 20 missing samples out of 100), then iterating through the roots would be `O(20)`, which seems bad to me. But perhaps getting a list of the roots in a tree is actually not such a performance hit?

**Comment by jeromekelleher** [comment]
Yes, it's the number of roots. But these arguments only matter if you have say 100K samples. If you only have 100 samples then it makes no difference at all how you compute things. Unless you are actually working with genuinely large trees there's no point in worrying about performance in terms of this stuff.

**Comment by jeromekelleher** [comment]
#386 would give an efficient solution to this.

**Comment by benjeffery** [comment]
Assuming that now #386 is complete that this is done.
